crates/red_knot_python_semantic/src/types/infer.rs

```diff
@@ -2469,7 +2477,24 @@ impl<'db> TypeInferenceBuilder<'db> {
             (UnaryOp::Invert, Type::BooleanLiteral(bool)) => Type::IntLiteral(!i64::from(bool)),
 
             (UnaryOp::Not, ty) => ty.bool(self.db).negate().into_type(self.db),
+            (_, Type::Any) => Type::Any,
+            (_, Type::Unknown) => Type::Unknown,
+            (op @ (UnaryOp::UAdd | UnaryOp::USub | UnaryOp::Invert), Type::Instance(class)) => {
+                let class_member = class.class_member(self.db, op.dunder());
+                let call = class_member.call(self.db, &[operand_type]);
+                if matches!(call, CallOutcome::NotCallable { not_callable_ty: _ }) {
```

Comment:
There is also the possibility of `CallOutcome::Union` where some elements of the union were not callable, in which case we should also emit this diagnostic. The way to handle this is to match on `call.return_ty_result(db, unary, self)`; if we get `Ok(return_ty)` we just return that and there is no diagnostic, if we get any `NotCallableError::Err` variant, we do emit the unsupported-operator diagnostic, and return the `return_ty` field of the `Err` variant.